server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java

```diff
@@ -1918,6 +1918,14 @@ public void forceMerge(final boolean flush, int maxNumSegments, boolean onlyExpu
                 }
                 if (flush) {
                     flush(false, true);
+
+                    // If any merges happened then we need to release the unmerged input segments so they can be deleted. A periodic refresh
+                    // will do this eventually unless the user has disabled refreshes or isn't searching this shard frequently, in which
+                    // case we should do something here to ensure a timely refresh occurs. However there's no real need to defer it nor to
+                    // have any should-we-actually-refresh-here logic: we're already doing an expensive force-merge operation at the user's
+                    // request and therefore don't expect any further writes so we may as well do the final refresh immediately and get it
+                    // out of the way.
+                    refresh("force-merge");
```

Comment:
In the case of anyone force merging against an index that still receives writes, I think this risks writing a new segment for the refresh. I did not look whether it is possible, but it would be ideal if we could somehow refresh to the latest available segment?

Reply:
I'm not sure I follow. Do you mean to refresh to the segments we just flushed, even if there are newer ones on disk? We don't necessarily flush exactly the result of the force-merge either, we're free to do more merges before flushing.

In principle we're fairly free to write out new segments and/or refresh whenever we decide it's appropriate. For instance we effectively refresh when moving a shard onto a different node, which may happen at any time.

Technically the Lucene API does let us reopen the underlying `DirectoryReader` to a specific index commit, but that's not something we ever do in Elasticsearch at the moment and I think it'd be excessively hard to do so here. We'd need to protect against things like undoing the effects of a concurrent refresh, for example.

Reply:
It was only to avoid writing out new segments, for instance if the index has refresh completely disabled.

But I thought the complexity would be somewhat like you described and I am good with how it is. Thanks for entertaining me!

Reply:
🕺

I see. Even if refresh is completely disabled we still write new segments to disk for lots of other reasons (e.g. if the indexing buffer or the translog gets too full). We just don't reopen the reader on the new segments as much, but we're not forbidden from doing so either.